File: service/internal/llm/gemini.go
package llm

import (
	"context"
	"fmt"
	"time"

	"google.golang.org/genai"
)

type GeminiClient struct {
	apiKey      string
	model       string
	temperature float64
	timeout     time.Duration
}

func NewGeminiClient(apiKey, model string, temperature float64, timeout int) *GeminiClient {
	return &GeminiClient{
		apiKey:      apiKey,
		model:       model,
		temperature: temperature,
		timeout:     time.Duration(timeout) * time.Second,
	}
}

func (c *GeminiClient) Generate(prompt string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	client, err := genai.NewClient(ctx, &genai.ClientConfig{APIKey: c.apiKey})
	if err != nil {
		return "", fmt.Errorf("failed to create Gemini client: %w", err)
	}

	// TODO Add temperature as option
	result, err := client.Models.GenerateContent(
		ctx,
		"gemini-2.5-flash",
		genai.Text(prompt),
		nil,
	)
	if err != nil {
		return "", fmt.Errorf("gemini generation failed: %w", err)
	}

	return result.Text(), nil
}

File: reviewer/llm/gemini.py
import asyncio
from google import genai
from google.genai import types

from llm.client import LlmClient


class GeminiClient(LlmClient):

    def __init__(self, api_key: str, model: str, temperature: float, timeout: int):
        self.api_key = api_key
        self.model = model
        self.temperature = temperature
        self.timeout = timeout
        self.client = genai.Client(api_key=api_key)

    async def generate(self, prompt: str) -> str:
        try:
            loop = asyncio.get_event_loop()
            response = await asyncio.wait_for(
                loop.run_in_executor(
                    None,
                    lambda: self.client.models.generate_content( # type: ignore
                        model=self.model,
                        contents=prompt,
                        config=types.GenerateContentConfig(
                            temperature=self.temperature
                        )
                    )
                ),
                timeout=self.timeout
            )
            
            return response.text or ""
        
        except asyncio.TimeoutError:
            raise Exception(f"Gemini generation timed out after {self.timeout} seconds")
        
        except Exception as error:
            raise Exception(f"Gemini generation failed: {error}")